### PR TITLE
discovery: pinned syncers

### PR DESCRIPTION
**Pull request by cfromknecht**
A pinned syncer is an ActiveSyncer that is configured to always remain
active for the lifetime of the connection. Pinned syncers do not count
towards the total NumActiveSyncer count, which are rotated periodically.

This features allows nodes to more tightly synchronize their routing
tables by ensuring they are always receiving gossip from distinguished
subset of peers.

### CODE REVIEW

File: config.go [hunk 1]
```diff
@@ -309,6 +309,8 @@ type Config struct {
 
 	Routing *lncfg.Routing `group:"routing" namespace:"routing"`
 
+	Gossip *lncfg.Gossip `group:"gossip" namespace:"gossip"`
```
Comment (wpaulino) [inline config.go:312]: Thoughts on moving the existing `numgraphsyncpeers` flag here and deprecating the top-level one while still parsing it? Then we can remove the top-level one at a later release.

File: lntest/itest/lnd_test.go [hunk 2]
```diff
@@ -10335,16 +10335,39 @@ func testGraphTopologyNotifications(net *lntest.NetworkHarness, t *harnessTest)
...
+	bob, err := net.NewNode("bob", nil)
+	require.NoError(t.t, err)
+	defer shutdownAndAssert(net, t, bob)
+
+	// Connect Alice and Bob.
+	ctxt, _ := context.WithTimeout(ctxb, defaultTimeout)
+	err = net.EnsureConnected(ctxt, alice, bob)
+	require.NoError(t.t, err)
+
+	// Alice stimmy.
```
Comment (wpaulino) [inline lntest/itest/lnd_test.go:10409]: :joy:

File: discovery/sync_manager.go [hunk 3]
```diff
@@ -140,6 +145,10 @@ type SyncManager struct {
 	// currently receiving new graph updates from.
 	inactiveSyncers map[route.Vertex]*GossipSyncer
 
+	pinnedSyncers map[route.Vertex]struct{}
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply discovery/sync_manager.go:148]: removed

File: discovery/sync_manager.go [hunk 4]
```diff
@@ -325,7 +326,9 @@ func (m *SyncManager) syncerHandler() {
 			// keep track of the corresponding syncer to properly
 			// handle disconnects. We'll also use a signal to know
 			// when the historical sync completed.
-			setHistoricalSyncer(s)
+			if !isPinnedSyncer {
+				setHistoricalSyncer(s)
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply discovery/sync_manager.go:330]: i tried going down that path but it gets messy, needing to select/transition between pinned vs unpinned peers to be the next historical syncer and test all the edge cases. it was much simpler just to treat them as separate.

File: discovery/sync_manager.go [hunk 5]
```diff
@@ -386,13 +388,15 @@ func (m *SyncManager) syncerHandler() {
 			// If we've already performed an initial historical
 			// sync, we won't try again with our passive peers.
 			if m.IsGraphSynced() {
+				m.cfg.HistoricalSyncTicker.Pause()
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply discovery/sync_manager.go:415]: removed the pauses in this area. the ticker is now only reset when setting the initial syncer. after that completes it remains active permanently.

File: discovery/sync_manager.go [hunk 6]
```diff
@@ -380,12 +380,16 @@ func (m *SyncManager) syncerHandler() {
 		// Our HistoricalSyncTicker has ticked, so we'll randomly select
 		// a peer and force a historical sync with them.
 		case <-m.cfg.HistoricalSyncTicker.Ticks():
-			s := m.forceHistoricalSync()
+			// If we've already performed an initial historical
+			// sync, we won't try again with our passive peers.
+			if m.IsGraphSynced() {
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply discovery/sync_manager.go:414]: Gotcha that makes sense, reversed the ordering

File: lnrpc/rpc.proto [hunk 7]
```diff
@@ -1487,6 +1487,13 @@ message Peer {
         Denotes that we are not receiving new graph updates from the peer.
         */
         PASSIVE_SYNC = 2;
+
+        /*
+        Denotes that this peer is pinned into an active sync.
+        */
+        PINNED_ACTIVE_SYNC = 3;
+
+        PINNED_PASSIVE_SYNC = 4;
```
[the remark this answers is not in the document]
Comment (cfromknecht) [reply lnrpc/rpc.proto:1496]: good call, ended up removing and now there's only `PINNED_SYNC`. we no longer need the two states with the introduction of `syncerIdle`

File: config.go (hunk 1 above)
Comment (cfromknecht) [reply config.go:312]: sgtm, let's do that in a follow up

**Review by wpaulino — APPROVED** [review]
LGTM. Looks like there's an flake (channel not opened before timeout) in one of the integration tests. I'm assuming it's the pre-existing one which @halseth is working on fixing with https://github.com/lightningnetwork/lnd/pull/4953, though it'd be nice to double check.

File: discovery/sync_manager.go [hunk 8]
```diff
@@ -214,9 +213,12 @@ func (m *SyncManager) syncerHandler() {
 		initialHistoricalSyncSignal chan struct{}
 	)
 
-	setHistoricalSyncer := func(s *GossipSyncer) {
+	setInitialHistoricalSyncer := func(s *GossipSyncer) {
 		initialHistoricalSyncer = s
 		initialHistoricalSyncSignal = s.ResetSyncedSignal()
+
+		m.cfg.HistoricalSyncTicker.Pause()
```
Comment (halseth) [inline discovery/sync_manager.go:230]: add a comment to why this is done?

File: discovery/sync_manager.go [hunk 9]
```diff
@@ -322,7 +323,9 @@ func (m *SyncManager) syncerHandler() {
 			// keep track of the corresponding syncer to properly
 			// handle disconnects. We'll also use a signal to know
 			// when the historical sync completed.
-			setInitialHistoricalSyncer(s)
+			if !isPinnedSyncer {
```
Comment (halseth) [inline discovery/sync_manager.go:326]: hm, why this condition?

File: lntest/itest/lnd_test.go [hunk 10]
```diff
@@ -10330,19 +10330,77 @@ func subscribeGraphNotifications(t *harnessTest, ctxb context.Context,
...
+			fmt.Sprintf("--gossip.pinned-syncers=%s",
+				bobInfo.IdentityPubkey),
+		}
+	}
+
+	alice, err := net.NewNode("alice", aliceArgs)
+	require.NoError(t.t, err)
+	defer shutdownAndAssert(net, t, alice)
+
+	// When testing pinnned syncing, we need to ensure Alice doesn't select
```
Comment (halseth) [inline lntest/itest/lnd_test.go:10386]: I don't get this. If Bob is the pinned peer, why do we need to check that Bob won't also be the active peer? Aren't we checking that Alice is syncing from dummy below, while we should be checking she syncs from Bob?

**Review by halseth — CHANGES_REQUESTED** [review]

File: lntest/itest/lnd_test.go (hunk 10 above)
Comment (cfromknecht) [reply lntest/itest/lnd_test.go:10386]: Alice will complete an initial sync with the dummy, which means her graph will also be empty (but `synced_to_graph` will show true).

This is mainly done to avoid false positives due to regressions, since the test will pass all the same if we are unintentionally in ACTIVE_SYNC with Bob.

File: discovery/sync_manager.go (hunk 9 above)
Comment (cfromknecht) [reply discovery/sync_manager.go:326]: Several reason, also mentioned in another comment w/ wilmer. 

Each pinned peer already does a historical sync, so there can be many of them at the same time. When it comes to replacement, ideally you'd prioritize pinned over passive syncers but the bookkeeping gets tricky when they're all happening in parallel. Disconnects need to be handled differently depending on pinned vs regular. Tl;dr the number of edge cases isn't worth it, and it's simpler to reason about them as separate peer groups.

My preferred approach if we are interested in the cosmetic aspect is just a callback that syncers can use to report they've completed a historical sync.

File: lntest/itest/lnd_test.go (hunk 10 above)
Comment (cfromknecht) [reply lntest/itest/lnd_test.go]: opted for more explicit check of the sync state, ptal

File: discovery/sync_manager.go [hunk 11]
```diff
@@ -140,6 +145,8 @@ type SyncManager struct {
 	// currently receiving new graph updates from.
 	inactiveSyncers map[route.Vertex]*GossipSyncer
 
+	pinnedActiveSyncers map[route.Vertex]*GossipSyncer
```
Comment (halseth) [inline discovery/sync_manager.go:148]: add a doc similar to other maps

File: discovery/sync_manager.go (hunk 9 above)
Comment (halseth) [reply discovery/sync_manager.go:331]: Yeah, I'm kinda worried that all these different states can become hard to reason about and maintain. Perhaps not in the scope of this PR, but it could use an overhaul

File: lncfg/gossip.go [hunk 12]
```diff
@@ -0,0 +1,30 @@
...
+}
+
+func (g *Gossip) Parse() error {
+	pinnedSyncers := make(discovery.PinnedSyncers)
+	for _, pubkeyStr := range g.PinnedSyncersRaw {
+		vertex, err := route.NewVertexFromStr(pubkeyStr)
+		if err != nil {
+			return err
+		}
+		fmt.Printf("vertex: %v\n", vertex)
```
Comment (halseth) [inline lncfg/gossip.go:23]: remove?

**Review by halseth — APPROVED** [review]
LGTM. Not sure if this should go into v0.12.1 though, as it is new functionality.

**Comment by cfromknecht** [comment]
addressed comments, also moved `markGraphSynced` into the syncer so that cosmetically we show `synced_to_graph` whenever any peer (pinned or regular) finishes a historical sync

**Comment by cfromknecht** [comment]
> Not sure if this should go into v0.12.1 though, as it is new functionality.

We have released new features in point releases before, main thing is that this change is not breaking since it is opt-in.

In this case, users have reported an inability to synchronize the routing tables of two directly-connected nodes. This occurs because when the number of open channels is high because the historical sync randomization takes a long time before selecting their node again.

**Review by halseth — APPROVED** [review]

**Review by wpaulino — APPROVED** [review]
